openfasoc/generators/glayout/glayout/flow/blocks/elementary/current_mirror/current_mirror.py

```diff
@@ -48,107 +48,209 @@ def current_mirror_netlist(
+        pdk,
+        interdigitized_fets.ports['A_gate_W'],
+        interdigitized_fets.ports['B_gate_W'],
+        extension=extension_length,
+        viaoffset=False
+    )
+
+    # Short gate and drain of the reference (FET A)
+    interdigitized_fets << L_route(
+        pdk,
+        interdigitized_fets.ports['A_drain_W'],
+        gate_short.ports['con_N'],
+        viaoffset=False,
+        fullbottom=False
+    )
+    return interdigitized_fets  
+
+
+def add_well_tie(pdk: MappedPDK, component: Component, interdigitized_fets_bbox, tie_layers: tuple[str, str], numcols: int) -> Component:
+    """Adds a well tie (tap ring) to the component.
+
+    Args:
+        interdigitized_fets_bbox: Bounding box of the interdigitized FETs to enclose.
+        tie_layers (tuple[str,str]): Layers for the tie ring.
+        numcols (int): Number of columns (used for port name calculation, fragile).
+
+    Returns:
+        Component: The component with the well tie added.
+    """
+    max_metal_separation = pdk.util_max_metal_seperation()
+    tap_sep = max(
+        max_metal_separation,
+        pdk.get_grule("active_diff", "active_tap")["min_separation"],
+    )
+    tap_sep += pdk.get_grule("p+s/d", "active_tap")["min_enclosure"]
+    tap_encloses = (
+        2 * (tap_sep + interdigitized_fets_bbox.xmax),
+        2 * (tap_sep + interdigitized_fets_bbox.ymax),
+    )
+    tie_ref = component << tapring(pdk, enclosed_rectangle=tap_encloses, sdlayer="p+s/d", horizontal_glayer=tie_layers[0], vertical_glayer=tie_layers[1])
+    component.add_ports(tie_ref.get_ports_list(), prefix="welltie_")
+
+
+    try: 
+        component << straight_route(pdk, component.ports["fet_A_0_dummy_L_gsdcon_top_met_W"], component.ports["welltie_W_top_met_W"], glayer2="met1")
+    except KeyError:
+        pass
+
+    try: 
+        end_col = numcols - 1
+        port1 = f'fet_B_{end_col}_dummy_R_gdscon_top_met_E'
+        component << straight_route(pdk, component.ports[port1], component.ports["welltie_E_top_met_E"], glayer2="met1")
+    except KeyError:
+        pass
+
+    return component 
+
+def add_substrate_tap(pdk: MappedPDK, component: Component, interdigitized_fets_bbox) -> Component:
+    """Adds a substrate tap ring to the component.
+
+    Args:
+        component (Component): The component to add the substrate tap to.
+    Returns:
+        Component: The component with the substrate tap added.
+    """
+    subtap_sep = pdk.get_grule("dnwell", "active_tap")["min_separation"]
+    subtap_enclosure = (
+        2.5 * (subtap_sep + interdigitized_fets_bbox.xmax), 
+        2.5 * (subtap_sep + interdigitized_fets_bbox.ymax), 
+    )
+    subtap_ring = component << tapring(pdk, enclosed_rectangle=subtap_enclosure, sdlayer="p+s/d", horizontal_glayer="met2", vertical_glayer="met1")
+    component.add_ports(subtap_ring.get_ports_list(), prefix="substrate_tap_")
+    return component
+
+
 def current_mirror(
```

Comment:
Have you tested if this function still works the same? For a temporary check, you can instantiate a few current mirrors and compare the results to the older function and ensure they are the same.

However, it would be really good if you add a test in the CI. You can take a look at [this](https://github.com/idea-fasoc/OpenFASOC/blob/main/.github/workflows/glayout_sky130.yml) file as an example.